src/lib/Scenes/Home/Components/ForYou/index.tsx

```diff
@@ -1,104 +1,124 @@
-    super(props)
-
-    const dataSource: ListViewDataSource = new ListView.DataSource({ rowHasChanged: (r1, r2) => r1 !== r2 })
-
-    this.state = {
-      isRefreshing: false,
-      modules: [],
-      dataSource,
-    }
+  state = {
+    isRefreshing: false,
+    rowData: [],
   }
 
   componentDidMount() {
-    const rows: DataSourceRow[] = []
-    const artworkModules = this.props.forYou.artwork_modules || []
-    const artistModules = this.props.forYou.artist_modules && this.props.forYou.artist_modules.concat()
-    const fairsModule = this.props.forYou.fairs_module
-    rows.push({ type: "fairs", data: fairsModule })
-    for (let i = 0; i < artworkModules.length; i++) {
-      const artworkModule = artworkModules[i]
-      rows.push({ type: "artwork", data: artworkModule })
-      if ((i + 1) % 2 === 0) {
+    const { forYou } = this.props
+    const rowData = []
+    const artworkModules = forYou.artwork_modules || []
+    const artistModules = forYou.artist_modules && forYou.artist_modules.concat()
+    const fairsModule = forYou.fairs_module
+
+    rowData.push({
+      type: "fairs",
+      data: fairsModule,
+    })
+
+    artworkModules.forEach((artworkModule, index) => {
+      rowData.push({
+        type: "artwork",
+        data: artworkModule,
+      })
+
+      const alternateRow = (index + 1) % 2 === 0
+
+      if (alternateRow) {
         const artistModule = artistModules.shift()
         if (artistModule) {
-          rows.push({ type: "artist", data: artistModule })
+          rowData.push({
+            type: "artist",
+            data: artistModule,
+          })
         }
       }
-    }
+    })
+
     this.setState({
-      dataSource: this.state.dataSource.cloneWithRows(rows),
+      rowData,
     })
   }
 
-  handleRefresh = () => {
+  handleRefresh = async () => {
     this.setState({ isRefreshing: true })
-    const stopRefreshing = () => this.setState({ isRefreshing: false })
-
-    Promise.all(
-      this.state.modules.map(module => {
-        return new Promise((resolve, reject) => {
-          module.forceFetch(null, readyState => {
-            if (readyState.error) {
-              reject(readyState.error)
-            } else if (readyState.aborted) {
-              reject()
-            } else if (readyState.done) {
-              resolve()
-            }
-          })
+
+    try {
+      await Promise.all(
+        this.railModules.map(railModule => {
+          railModule.refreshData()
         })
+      )
+
+      this.setState({
+        isRefreshing: false,
+      })
+    } catch (error) {
+      console.error("ForYou/index.tsx", error.message)
+
+      this.setState({
+        error: error.message, // TODO: Display this somehow
       })
-    ).then(stopRefreshing, stopRefreshing)
+    }
+  }
+
+  registerRailModule = row => {
+    return railModule => {
+      this.railModules[row - 2] = railModule // Offset row because we don’t store a reference to the search bar and hero units rows.
```

Comment:
Out of curiosity I tried replacing this line with `this.railModules.push(railModule)` and things seemed to work no differently? Is this maybe out of date?

Reply:
Right, I'm not sure what the reasoning was for `row - 2` but it seems to be outdated. I think it's safe to replace with `push` instead

Reply:
No this is because different list-y views have different ways of how they manage their internal components. We should _not_ put any assumptions on these references being yielded only once and in the order we expect, rather just replace whatever was at the exact we want the reference to be.

Reply:
Ah ok, got it. Thanks for the added context @alloy

Reply:
👍